### PR TITLE
Harmonize data streams

### PR DESCRIPTION
**Pull request by kristinksellers**
Huge pull request - apologies for how much different functionality is included.

- Stimulation settings from both stimLog and deviceSettings
- Harmonizing derivedTimes across data streams (using time domain as 'base time')
- Pulling in additional meta data
- Pulling in adaptive settings
- Bug fixes to address edge case recordings
- Pulling in adaptive time series
- Handing files where power / FFT / adaptive channels have change in sampling rate

Documentation update pending

### CODE REVIEW

**Comment by roeegilron** [comment]
So much to like in this pull request! It works much better on longer recordings as discussed and fixed a lot of the previous edge cases we found

**Review by roeegilron — COMMENTED** [review]
This is good! Did not check all the stim /sense settings.

**Comment by roeegilron** [comment]
in `deserializeJSON` if you have a failure in the `try` block you don't return an empty `data` variable

**Comment by roeegilron** [comment]
> in `deserializeJSON` if you have a failure in the `try` block you don't return an empty `data` variable

to be clear should be like:

`try
    data = [];
    data = json.load(filename);
    fprintf('File loaded in %.2f seconds\n',toc(start));
catch
`

**Comment by kristinksellers** [comment]
@roeegilron See b6f5a51dc496652aec4e03fc1e18a6e9f8bb6049 for update to `deserializeJSON` to return empty `data` variable

File: code/assignTime.m [hunk 1]
```diff
@@ -64,12 +64,28 @@
...
+% trying to confirm there is enough time to assign to samples without
+% causing overlap.
+% Remove at most 6 adjacent packets (to prevent large un-needed
 % packet rejection driven by positive outliers)
+numPackets = size(dataTable_original,1);
 indices_backInTime = [];
 for iIndex = 1:length(diffIndices)
-    counter = 1;
-    while (counter < 6) && dataTable_original.PacketGenTime(diffIndices(iIndex) + counter)...
+    counter = 3;
```
Comment (jansoromeo) [inline code/assignTime.m:75]: may be add to the right of this line why counter = 3. I guess it has to do with the [diffIndices + 2]

File: code/assignTime.m [hunk 2]
```diff
@@ -162,25 +178,73 @@
...
 % Loop through each chunk to determine offset to apply (as determined by
-% average difference between packetGenTime and expectedElapsed)
+% average difference between packetGenTime and expectedElapsed) --
+% calculated for all chunks here, will subsequently only apply error for
+% chunks which are preceeded by gaps >= 6 seconds
 disp('Determining start time of each chunk')
 
 % PacketGenTime in ms; convert difference to 1e-4 seconds, units of
 % systemTick and expectedElapsed
 diff_PacketGenTime = [1; diff(dataTable.PacketGenTime) * 1e1];
```
Comment (jansoromeo) [inline code/assignTime.m:234]: it is unclear why multiplying by 1e1 (10) - may be add a comment to the right?

File: code/assignTime.m (hunk 1 above)
Comment (kristinksellers) [reply code/assignTime.m]: Resolved with 83db53bba4ad68982cbfeac256da13d0034a9baa

File: code/assignTime.m (hunk 2 above)
Comment (kristinksellers) [reply code/assignTime.m]: Resolved with 83db53bba4ad68982cbfeac256da13d0034a9baa

**Pull request merged**